### PR TITLE
AIP006, AIP007 and EDVerify precompiled contract implementation

### PR DESCRIPTION
**Pull request by AionJayT**
## Notice

It is not allowed to submit your PR to the master branch directly, please submit your PR to the master-pre-merge branch.

## Description

Please include a brief summary of the change that this pull request proposes. Include any relevant motivation and context. List any dependencies required for this change.

This PR is merged the EDVerify implementation by Centrys team [APB-46](https://github.com/aionnetwork/aion/pull/589) and implemented the Blake2bHash, TransactionHash precompiled contract by [AIP006](https://github.com/aionnetwork/AIP/issues/8), [AIP007](https://github.com/aionnetwork/AIP/issues/10). This PR requires the solidity and fastVM native libraries update (v0.3.2). Please see the related PR in the fastVM repo [PR#29](https://github.com/aionnetwork/aion_fastvm/pull/29)

Note:
This PR will break the consensus (hard fork). Need to communicate with the Aion community to upgrade their node when the new binary release. 

Fixes Issue # .

## Type of change

Insert **x** into the following checkboxes to confirm (eg. [x]):
- [ ] Bug fix.
- [x] New feature.
- [ ] Enhancement.
- [x] Unit test.
- [x] Breaking change (a fix or feature that causes existing functionality to not work as expected).
- [x] Requires documentation update.

## Testing

Please describe the tests you used to validate this pull request. Provide any relevant details for test configurations as well as any instructions to reproduce these results.

-

## Verification

Insert **x** into the following checkboxes to confirm (eg. [x]):
- [x] I have self-reviewed my own code and conformed to the style guidelines of this project.
- [x] New and existing tests pass locally with my changes.
- [x] I have added tests for my fix or feature.
- [ ] I have made appropriate changes to the corresponding documentation.
- [x] My code generates no new warnings.
- [x] Any dependent changes have been made.

### CODE REVIEW

File: modPrecompiled/src/org/aion/precompiled/contracts/Blake2bHashContract.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+            return new ExecutionResult(ResultCode.OUT_OF_NRG, 0);
+        }
+
+        // check operation number
+//        int operation = input[0];
+//
+//        switch (operation) {
+//            case 0:
+        return blake256Hash(input, nrgLeft);
+//            case 1:
```
Comment (aion-kelvin) [inline modPrecompiled/src/org/aion/precompiled/contracts/Blake2bHashContract.java]: delete commented stuff if not needed

File: modPrecompiled/src/org/aion/precompiled/contracts/Blake2bHashContract.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    private ExecutionResult blake256Hash(byte[] input, long nrg) {
+        byte[] hash = blake256(input);
+        return new ExecutionResult(ResultCode.SUCCESS, nrg, hash);
+    }
+
+//    private ExecutionResult blake128Hash(byte[] input, long nrg) {
+//        byte[] byteArray = new byte[input.length - 1];
+//        System.arraycopy(input, 1, byteArray, 0, input.length - 1);
```
Comment (aion-kelvin) [inline modPrecompiled/src/org/aion/precompiled/contracts/Blake2bHashContract.java]: delete commented stuff if not needed

File: modPrecompiled/src/org/aion/precompiled/contracts/Blake2bHashContract.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+     */
+    public ExecutionResult execute(byte[] input, long nrg) {
+
+        // check length
+        if (input == null || input.length == 0 || input.length > 1_048_576L) {
+            return new ExecutionResult(
+                ResultCode.INTERNAL_ERROR, nrg - COST, INPUT_LENGTH_ERROR_MESSAGE.getBytes());
+        }
+
+        long additionalNRG = ((long) Math.ceil(((double) input.length - 1) / WORD_LENGTH)) * 6;
```
Comment (aion-kelvin) [inline modPrecompiled/src/org/aion/precompiled/contracts/Blake2bHashContract.java]: out of curiosity, how was this formula determined?

File: modPrecompiled/src/org/aion/precompiled/contracts/ContractBase.java [hunk 4]
```diff
@@ -0,0 +1,5 @@
+package org.aion.precompiled.contracts;
+
+public class ContractBase {
```
Comment (aion-kelvin) [inline modPrecompiled/src/org/aion/precompiled/contracts/ContractBase.java]: is this file needed?

File: modPrecompiled/src/org/aion/precompiled/contracts/EDVerifyContract.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+    // set to a default cost for now, this will need to be adjusted
+    private final static long COST = 3700L;
+
+    @Override
+    public IExecutionResult execute(byte[] input, long nrgLimit) {
+        if (COST > nrgLimit) {
+            return new ExecutionResult(ExecutionResult.ResultCode.OUT_OF_NRG, 0);
+        }
+        byte[] msg = new byte[32];
```
Comment (aion-kelvin) [inline modPrecompiled/src/org/aion/precompiled/contracts/EDVerifyContract.java]: would it be worthwhile to validate the input size of the 'input' array first?  that way we can return an ExecutionResult with an error

File: modPrecompiled/src/org/aion/precompiled/contracts/Blake2bHashContract.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+     * <p>the returned hash is in ExecutionResult.getOutput
+     *
+     * <p>the maximum data hash is 1M bytes
+     */
+    public ExecutionResult execute(byte[] input, long nrg) {
+
+        // check length
+        if (input == null || input.length == 0 || input.length > 1_048_576L) {
+            return new ExecutionResult(
+                ResultCode.INTERNAL_ERROR, nrg - COST, INPUT_LENGTH_ERROR_MESSAGE.getBytes());
```
Comment (aion-kelvin) [inline modPrecompiled/src/org/aion/precompiled/contracts/Blake2bHashContract.java]: would ResultCode.FAILURE make more sense?  INTERNAL_ERROR makes it sound like there was a bug in the contract or VM, when the real problem is that the user specified invalid input

File: modPrecompiled/src/org/aion/precompiled/contracts/EDVerifyContract.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+        byte[] sig = new byte[64];
+        byte[] pubKey = new byte[32];
+
+        System.arraycopy(input, 0, msg, 0, 32);
+        System.arraycopy(input, 32, sig, 0, 64);
+        System.arraycopy(input, 96, pubKey, 0, 32);
+
+        //TODO: may need to remove after end-to-end test
```
Comment (aion-kelvin) [inline modPrecompiled/src/org/aion/precompiled/contracts/EDVerifyContract.java]: can we remove these now?  or is the end-to-end test not happened yet?

File: modPrecompiled/test/org/aion/precompiled/contracts/EDVerifyContractTest.java [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+        System.arraycopy(pubKey, 0, input, 96, 32);
+
+        ExecutionContext ctx = new ExecutionContext(txHash,
+            ContractFactory.getEdVerifyContractAddress(), origin, caller, nrgPrice,
+            nrgLimit, callValue,
+            callData, depth, kind, flags, blockCoinbase, blockNumber, blockTimestamp, blockNrgLimit,
+            blockDifficulty);
+        IPrecompiledContract contract = new ContractFactory().getPrecompiledContract(ctx, null);
+
+        assert contract != null;
```
Comment (aion-kelvin) [inline modPrecompiled/test/org/aion/precompiled/contracts/EDVerifyContractTest.java]: can we use assert() or assertThat() instead ?

File: build.gradle [hunk 9]
```diff
@@ -175,7 +175,11 @@ task downloadFastVmGit(type:Download) {
     dest buildDir
 }
 
-task downloadSolidityGit(type:Download) { 
+task downloadSolidityGit(type:Download) {
+    doFirst {
+        assert file("./solidity_v${vmVersion}.tar.gz").exists()
```
[the remark this answers is not in the document]
Comment (AionJayT) [reply build.gradle]: oh. this will remove

File: build.gradle [hunk 10]
```diff
@@ -192,8 +196,8 @@ task getVmDependencies {
     } else {
         // 'overwrite' property of Download task doesn't seem to work when
         // there's multiple files in the task, so splitting them into two
-        dependsOn downloadFastVmGit
-        dependsOn downloadSolidityGit
+        //dependsOn downloadFastVmGit
```
[the remark this answers is not in the document]
Comment (AionJayT) [reply build.gradle]: comment it cause the current vm native builds not release yet. will rollback when vm native lib has been released.

File: modPrecompiled/src/org/aion/precompiled/ContractFactory.java [hunk 11]
```diff
@@ -28,23 +28,42 @@
...
+    private static final String ADDR_ED_VERIFY =
+        "0000000000000000000000000000000000000000000000000000000000000010";
+    private static final String ADDR_BLAKE2B_HASH =
+        "0000000000000000000000000000000000000000000000000000000000000011";
+    private static final String ADDR_TX_HASH =
+        "0000000000000000000000000000000000000000000000000000000000000012";
 
-    public ContractFactory() {}
+    private static IPrecompiledContract PC_ED_VERIFY;
+    private static IPrecompiledContract PC_BLAKE2B_HASH;
```
[the remark this answers is not in the document]
Comment (AionJayT) [reply modPrecompiled/src/org/aion/precompiled/ContractFactory.java]: Yes, just think these precompiled contract will be called a lot later on. So I put static on it.

File: modPrecompiled/src/org/aion/precompiled/ContractFactory.java [hunk 12]
```diff
@@ -28,23 +28,42 @@
...
+
+    public ContractFactory() {
+        PC_ED_VERIFY = new EDVerifyContract();
+        PC_BLAKE2B_HASH = new Blake2bHashContract();
```
[the remark this answers is not in the document]
Comment (AionJayT) [reply modPrecompiled/src/org/aion/precompiled/ContractFactory.java]: okay. will do a static {}

File: modPrecompiled/src/org/aion/precompiled/ContractFactory.java [hunk 13]
```diff
@@ -88,9 +115,12 @@ public IPrecompiledContract getPrecompiledContract(
      */
     public static boolean isPrecompiledContract(Address address) {
         switch (address.toString()) {
-            case TOTAL_CURRENCY:
-                return true;
-            case TOKEN_BRIDGE:
+            case ADDR_TOTAL_CURRENCY:
+                return false;
```
[the remark this answers is not in the document]
Comment (AionJayT) [reply modPrecompiled/src/org/aion/precompiled/ContractFactory.java]: okay, will change it.

File: modPrecompiled/src/org/aion/precompiled/contracts/ContractBase.java (hunk 4 above)
Comment (AionJayT) [reply modPrecompiled/src/org/aion/precompiled/contracts/ContractBase.java]: cause I put a static method originally from to bridge contract to here, there are some contracts can use the same statics method but in the different package.

File: modPrecompiled/src/org/aion/precompiled/contracts/Blake2bHashContract.java (hunk 6 above)
Comment (AionJayT) [reply modPrecompiled/src/org/aion/precompiled/contracts/Blake2bHashContract.java]: will change it.

File: modPrecompiled/src/org/aion/precompiled/contracts/EDVerifyContract.java (hunk 7 above)
Comment (AionJayT) [reply modPrecompiled/src/org/aion/precompiled/contracts/EDVerifyContract.java]: sure. I think that's it. The centrys doesn't update the implementation

File: modPrecompiled/test/org/aion/precompiled/contracts/EDVerifyContractTest.java (hunk 8 above)
Comment (AionJayT) [reply modPrecompiled/test/org/aion/precompiled/contracts/EDVerifyContractTest.java]: sure. it's suggested by IDE.

File: modPrecompiled/src/org/aion/precompiled/contracts/Blake2bHashContract.java (hunk 3 above)
Comment (AionJayT) [reply modPrecompiled/src/org/aion/precompiled/contracts/Blake2bHashContract.java]: follow by the Ethereum hash precompiledContract charge.

File: modPrecompiled/src/org/aion/precompiled/ContractFactory.java [hunk 14]
```diff
@@ -28,23 +28,45 @@
...
+import org.aion.precompiled.contracts.EDVerifyContract;
+import org.aion.precompiled.contracts.TXHashContract;
 import org.aion.vm.ExecutionContext;
 import org.aion.vm.IContractFactory;
 import org.aion.vm.IPrecompiledContract;
 
-/** A factory class that produces pre-compiled contract instances. */
+/**
+ * A factory class that produces pre-compiled contract instances.
+ */
```
Comment (AlexandraRoatis) [inline modPrecompiled/src/org/aion/precompiled/ContractFactory.java]: formatting

File: modPrecompiled/src/org/aion/precompiled/ContractFactory.java [hunk 15]
```diff
@@ -28,23 +28,45 @@
...
+    private static IPrecompiledContract PC_BLAKE2B_HASH;
+
+    static {
+        PC_ED_VERIFY = new EDVerifyContract();
+        PC_BLAKE2B_HASH = new Blake2bHashContract();
+    }
 
-    public ContractFactory() {}
+    public ContractFactory() {
+    }
```
Comment (AlexandraRoatis) [inline modPrecompiled/src/org/aion/precompiled/ContractFactory.java]: formatting

File: modPrecompiled/src/org/aion/precompiled/ContractFactory.java [hunk 16]
```diff
@@ -56,25 +78,33 @@ public ContractFactory() {}
...
+                    return null;
+                }
 
                 return contract;
+            case ADDR_ED_VERIFY:
+                return PC_ED_VERIFY;
+            case ADDR_BLAKE2B_HASH:
+                return PC_BLAKE2B_HASH;
+            case ADDR_TX_HASH:
+                return new TXHashContract(context);
```
Comment (AlexandraRoatis) [inline modPrecompiled/src/org/aion/precompiled/ContractFactory.java]: what's the reasoning behind handling the TX contract differently?

File: modPrecompiled/src/org/aion/precompiled/ContractFactory.java [hunk 17]
```diff
@@ -103,6 +135,25 @@ public static boolean isPrecompiledContract(Address address) {
...
+
+    /**
+     * Returns the address of the TxHash contract.
+     *
+     * @return the contract address
+     */
+    public static Address getTxHashContractAddress() {
+        return Address.wrap(ADDR_TX_HASH);
+    }
+
```
Comment (AlexandraRoatis) [inline modPrecompiled/src/org/aion/precompiled/ContractFactory.java]: shouldn't a getter for the blacke2b address be added here as well?

File: modPrecompiled/src/org/aion/precompiled/contracts/ContractBase.java [hunk 18]
```diff
@@ -0,0 +1,5 @@
+package org.aion.precompiled.contracts;
```
Comment (AlexandraRoatis) [inline modPrecompiled/src/org/aion/precompiled/contracts/ContractBase.java]: missing the license (in case the class is kept)

File: modPrecompiled/src/org/aion/precompiled/contracts/TXHashContract.java [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+import org.aion.vm.ExecutionContext;
+import org.aion.vm.ExecutionResult;
+import org.aion.vm.IPrecompiledContract;
+
+public class TXHashContract implements IPrecompiledContract {
+    public static final long COST = 500L;
+
+    private final ExecutionContext context;
+
+    public TXHashContract(@Nonnull final ExecutionContext _context) {
```
Comment (AlexandraRoatis) [inline modPrecompiled/src/org/aion/precompiled/contracts/TXHashContract.java]: there's no actual standard for non-null annotations: https://stackoverflow.com/questions/35892063/which-nonnull-java-annotation-to-use; I think null checks would be a better choice here

File: modPrecompiled/src/org/aion/precompiled/contracts/Blake2bHashContract.java (hunk 3 above)
Comment (yulongaion) [reply modPrecompiled/src/org/aion/precompiled/contracts/Blake2bHashContract.java]: Since blake2b is less computation intensive, should we decrease the nrg cost?

File: modVM/src/org/aion/vm/ExecutionContext.java [hunk 20]
```diff
@@ -120,6 +121,8 @@ public ExecutionContext(
         this.blockNrgLimit = blockNrgLimit;
         this.blockDifficulty = blockDifficulty;
         this.txHash = txHash;
+        this.originalTxHash = txHash;
```
Comment (yulongaion) [inline modVM/src/org/aion/vm/ExecutionContext.java]: The `txHash` is actually the original transaction hash. Suggest removing one of the them.

File: modPrecompiled/src/org/aion/precompiled/ContractFactory.java [hunk 21]
```diff
@@ -88,10 +118,12 @@ public IPrecompiledContract getPrecompiledContract(
...
         switch (address.toString()) {
-            case TOTAL_CURRENCY:
-                return true;
-            case TOKEN_BRIDGE:
+            case ADDR_TOKEN_BRIDGE:
+            case ADDR_ED_VERIFY:
+            case ADDR_BLAKE2B_HASH:
+            case ADDR_TX_HASH:
                 return true;
+            case ADDR_TOTAL_CURRENCY:
```
[the remark this answers is not in the document]
Comment (AionJayT) [reply modPrecompiled/src/org/aion/precompiled/ContractFactory.java]: Yao's opinion : yes!

File: modPrecompiled/src/org/aion/precompiled/contracts/Blake2bHashContract.java [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+     * Returns the hash of given input
+     *
+     * @param input data input; must be less or equal than 1 MB
+     *
+     * @return the returned blake2b 256bits hash is in ExecutionResult.getOutput
+     */
+    public ExecutionResult execute(byte[] input, long nrg) {
+
+        // check length
+        if (input == null || input.length == 0 || input.length > 1_048_576L) {
```
[the remark this answers is not in the document]
Comment (AionJayT) [reply modPrecompiled/src/org/aion/precompiled/contracts/Blake2bHashContract.java]: We should discuss within AIP. The basic idea is the transaction nrg limit is 2M. base on the gas cost of the precompiled contract in Ethereum, the formula is :  input divided by a word size (4 bytes) and times 6 plus base cost 30, approximate  1.5 times the input length.

File: modPrecompiled/src/org/aion/precompiled/contracts/Blake2bHashContract.java (hunk 3 above)
Comment (AionJayT) [reply modPrecompiled/src/org/aion/precompiled/contracts/Blake2bHashContract.java]: we should, but how do we decide the cost?  @yulongaion do you have any idea? I will try 3to run some benchmark compare with sha3.

File: modPrecompiled/src/org/aion/precompiled/contracts/ContractBase.java (hunk 18 above)
Comment (AionJayT) [reply modPrecompiled/src/org/aion/precompiled/contracts/ContractBase.java]: remove this class.

File: modVM/src/org/aion/vm/ExecutionContext.java (hunk 20 above)
Comment (AionJayT) [reply modVM/src/org/aion/vm/ExecutionContext.java]: @yulongaion not really, when the VM executing the internal tx, the txHash will be modified to the hash of the internal tx.

File: modPrecompiled/src/org/aion/precompiled/contracts/EDVerifyContract.java (hunk 5 above)
Comment (AionJayT) [reply modPrecompiled/src/org/aion/precompiled/contracts/EDVerifyContract.java]: check added

File: modPrecompiled/src/org/aion/precompiled/ContractFactory.java (hunk 16 above)
Comment (AionJayT) [reply modPrecompiled/src/org/aion/precompiled/ContractFactory.java]: cause the TxHash need the context setup.

File: modPrecompiled/src/org/aion/precompiled/ContractFactory.java (hunk 17 above)
Comment (AionJayT) [reply modPrecompiled/src/org/aion/precompiled/ContractFactory.java]: added

File: modPrecompiled/src/org/aion/precompiled/ContractFactory.java (hunk 21 above)
[the remark this answers is not in the document]
Comment (AionJayT) [reply modPrecompiled/src/org/aion/precompiled/ContractFactory.java]: I will discuss with Yao. then decide it

File: modPrecompiled/src/org/aion/precompiled/contracts/TXHashContract.java (hunk 19 above)
Comment (AionJayT) [reply modPrecompiled/src/org/aion/precompiled/contracts/TXHashContract.java]: the context should always non-null when constructing the Class. Adding this mainly tell the compiler the input should non-null.

File: modP2pImpl/src/org/aion/p2p/impl1/tasks/TaskWrite.java [hunk 23]
```diff
@@ -60,7 +60,7 @@ public void run() {
         if (channelBuffer.isClosed()) {
             channelBuffer.refreshHeader();
             channelBuffer.refreshBody();
-            p2pMgr.dropActive(channelBuffer.getNodeIdHash(), "close-already");
+            p2pMgr. dropActive(channelBuffer.getNodeIdHash(), "close-already");
```
Comment (AlexandraRoatis) [inline modP2pImpl/src/org/aion/p2p/impl1/tasks/TaskWrite.java]: accidental space added
